src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java

```diff
@@ -108,4 +150,73 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     protected String parserName() {
         return IdsQueryParser.NAME;
     }
+
+    public Query toQuery(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        if (this.ids.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+
+        Collection<String> typesForQuery = this.types;
+        if (typesForQuery == null || typesForQuery.isEmpty()) {
+            typesForQuery = parseContext.queryTypes();
+        } else if (typesForQuery.size() == 1 && Iterables.getFirst(typesForQuery, null).equals("_all")) {
+            typesForQuery = parseContext.mapperService().types();
+        }
+
+        List<BytesRef> ids = new ArrayList<>(this.ids.size());
+        for (String value : this.ids) {
+            BytesRef ref = new BytesRef(value);
+            ids.add(ref);
+        }
+
+        TermsQuery query = new TermsQuery(UidFieldMapper.NAME, Uid.createTypeUids(typesForQuery, ids));
+        query.setBoost(boost);
+        if (queryName != null) {
+            parseContext.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        // all fields can be empty or null
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        this.types = (List<String>) in.readGenericValue();
+        this.ids = (List<String>) in.readGenericValue();
```

Comment:
You mean re-adding the loop just for reading? read/writeGenericValue in the StreamInput/Output does support List, so except for the cast I don't understand the difference when I either use the loops as before or add a special read/write for List<String> to the Stream classes. Or maybe I don't understand.